src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/Common/SqlRetryLogicProvider.cs

```diff
@@ -0,0 +1,227 @@
+                    exceptions.Add(e);
+                    if (retryLogic.TryNextInterval(out TimeSpan intervalTime))
+                    {
+                        // The retrying event raises on each retry.
+                        ApplyRetryingEvent(sender, retryLogic, intervalTime, exceptions, e);
+
+                        await Task.Delay(intervalTime, cancellationToken);
+                        goto retry;
+                    }
+                    else
+                    {
+                        throw CreateException(exceptions, retryLogic);
+                    }
+                }
+                else
+                {
+                    RetryLogicPoolAdd(retryLogic);
+                    throw;
+                }
+            }
+        }
+
+        /// <summary>Executes a function and applies the retry logic if it's enabled.</summary>
+        public override async Task ExecuteAsync(object sender, Func<Task> function, CancellationToken cancellationToken = default)
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException(nameof(function));
+            }
+
+            SqlRetryLogicBase retryLogic = null;
+            var exceptions = new List<Exception>();
+        retry:
+            try
+            {
+                await function.Invoke();
+                RetryLogicPoolAdd(retryLogic);
+            }
+            catch (Exception e)
+            {
+                if (enableRetryLogic && RetryLogic.RetryCondition(sender) && RetryLogic.TransientPredicate(e))
+                {
+                    retryLogic = retryLogic ?? GetRetryLogic();
+                    SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.ExecuteAsync|INFO> Found the action eligible to apply the retry policy (retried attempts = {1}).", _typeName, retryLogic.Current);
+                    exceptions.Add(e);
+                    if (retryLogic.TryNextInterval(out TimeSpan intervalTime))
+                    {
+                        // The retrying event raises on each retry.
+                        ApplyRetryingEvent(sender, retryLogic, intervalTime, exceptions, e);
+
+                        await Task.Delay(intervalTime, cancellationToken);
+                        goto retry;
+                    }
+                    else
+                    {
+                        throw CreateException(exceptions, retryLogic);
+                    }
+                }
+                else
+                {
+                    RetryLogicPoolAdd(retryLogic);
+                    throw;
+                }
+            }
+        }
+
+        #region private methods
+
+        private Exception CreateException(IList<Exception> exceptions, SqlRetryLogicBase retryLogic, bool manualCancellation = false)
+        {
+            string message;
+            if (manualCancellation)
+            {
+                message = StringsHelper.GetString(Strings.SqlRetryLogic_RetryCanceled, retryLogic.Current);
```

Comment:
What is the type of this exception?
Logical exceptions (when logical decisions are made by driver) should be thrown as "SqlException" type.

Also, please use a utility like ADP to read strings from String Builder and call Utility methods here.